### PR TITLE
Add Kafka Producer metrics for send record failures

### PR DESCRIPTION
**Pull request by dinujoh**
### Description
Add Kafka Producer metrics for send record failures
 
### Check List
- [x] New functionality includes testing.
- [ ] New functionality has a documentation issue. Please link to it in this PR.
  - [ ] New functionality has javadoc added
- [x] Commits are signed with a real name per the DCO

By submitting this pull request, I confirm that my contribution is made under the terms of the Apache 2.0 license.
For more information on following Developer Certificate of Origin and signing off your commits, please check [here](https://github.com/opensearch-project/data-prepper/blob/main/CONTRIBUTING.md).

### CODE REVIEW

**Review by graytaylor0 — DISMISSED** [review]

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java [hunk 1]
```diff
@@ -183,7 +184,8 @@ public boolean validateSchema(final String jsonData, final String schemaJson) th
     private Callback callBack(final Object dataForDlq) {
         return (metadata, exception) -> {
             if (null != exception) {
-                LOG.error("Error occured while publishing " + exception.getMessage());
+                LOG.error("Error occurred while publishing " + exception.getMessage());
```
Comment (dlvenable) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java:187]: Please update this line to use SLF4J's string interpolation:

```
LOG.error("Error occurred while publishing ", exception.getMessage());
```

**Review by dlvenable — CHANGES_REQUESTED** [review]

**Review by dlvenable — DISMISSED** [review]

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java (hunk 1 above)
Comment (dinujoh) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java]: Fixed

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java [hunk 2]
```diff
@@ -102,8 +103,8 @@ public void produceRawData(final byte[] bytes, final String key) {
             send(topicName, key, bytes).get();
             topicMetrics.update(producer);
         } catch (Exception e) {
-            // TODO: Metrics
-            LOG.error("Error occurred while publishing {}", e.getMessage());
+            topicMetrics.getNumberOfRawDataSendErrors().increment();
```
Comment (hshardeesi) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java:106]: nit: RawDataSend does not provide good indication on what this counter is about.

**Review by graytaylor0 — APPROVED** [review]

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java [hunk 3]
```diff
@@ -183,7 +184,8 @@ public boolean validateSchema(final String jsonData, final String schemaJson) th
     private Callback callBack(final Object dataForDlq) {
         return (metadata, exception) -> {
             if (null != exception) {
-                LOG.error("Error occured while publishing " + exception.getMessage());
+                LOG.error("Error occurred while publishing {}", exception.getMessage());
+                topicMetrics.getNumberOfRecordProcessingException().increment();
```
Comment (hshardeesi) [inline data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java:188]: ...ReccordProcessing**Errors**?

**Review by hshardeesi — APPROVED** [review]

File: data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java (hunk 3 above)
Comment (dinujoh) [reply data-prepper-plugins/kafka-plugins/src/main/java/org/opensearch/dataprepper/plugins/kafka/producer/KafkaCustomProducer.java]: done

**Review by engechas — APPROVED** [review]

**Review by dlvenable — APPROVED** [review]

**Review by kkondaka — APPROVED** [review]

**Pull request merged**